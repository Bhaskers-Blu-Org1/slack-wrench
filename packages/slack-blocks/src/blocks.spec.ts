import { KnownBlock } from '@slack/types';
import { chain } from 'ramda';

import {
  Actions,
  Button,
  Context,
  DateString,
  Divider,
  Markdown,
  MdSection,
  PlainText,
  Section,
} from './blocks';

describe('Slack Block widgets', () => {
  const text = 'Ravenclaw is obviously the best house.';
  const buttonId = 'ravenAgree';

  it('renders a divider', () => {
    expect.assertions(1);
    expect(Divider()).toMatchSnapshot();
  });

  it('renders markdown text', () => {
    expect.assertions(1);
    expect(Markdown(text)).toMatchSnapshot();
  });

  it('renders a plaintext block', () => {
    expect.assertions(1);
    expect(PlainText(text)).toMatchSnapshot();
  });

  it('renders a plaintext block without emoji', () => {
    expect.assertions(1);
    expect(PlainText(text, false)).toMatchSnapshot();
  });

  it('renders a section', () => {
    expect.assertions(1);
    expect(
      Section({
        text: PlainText(text),
        fields: [PlainText(text), Markdown(text)],
        accessory: Button(text, buttonId),
      }),
    ).toMatchSnapshot();
  });

  it('renders a markdown section', () => {
    expect.assertions(1);
    expect(MdSection(text)).toMatchSnapshot();
  });

  it('renders an actions', () => {
    expect.assertions(1);
    expect(
      Actions([Button(text, buttonId), Button(text, `${buttonId}-2`)]),
    ).toMatchSnapshot();
  });

  it('renders a button', () => {
    expect.assertions(1);
    expect(Button(text, buttonId)).toMatchSnapshot();
  });

  it('renders context', () => {
    expect.assertions(1);
    expect(Context([PlainText(text)])).toMatchSnapshot();
  });

  it('creates a formatted date', () => {
    expect.assertions(1);
    expect(
      DateString(1567191506, 'Posted {date_num}', 'Posted 2019-09-30'),
    ).toMatchSnapshot();
  });

  it('can compose blocks into bigger things', () => {
    expect.assertions(1);

    const channels = [
      {
        name: 'house-ravenclaw',
        purpose: 'Discuss Ravenclaw business',
        lastPost: 1575643433,
        memberCount: 120,
      },
      {
        name: 'house-hufflepuff',
        purpose: 'Discuss Hufflepuff puffery',
        memberCount: 107,
        lastPost: 1575643433,
      },
    ];

    expect([
      MdSection(`Let me help you find some channels.`, {
        accessory: Button('Search', 'changeSearch'),
      }),
      Divider(),
      MdSection('*Channels*'),
      ...chain(
        (channel: any): KnownBlock[] => [
          MdSection(`*${channel.name}*\n${channel.purpose}`),
          Context([
            Markdown(
              `${channel.memberCount} members\n` +
                `Last post: ${DateString(
                  channel.lastPost,
                  '{date_pretty}',
                  channel.lastPost.toString(),
                )}`,
            ),
          ]),
          Actions([
            Button(':thumbsup:', 'thumbsUp', {
              value: channel.name,
            }),
            Button(':thumbsdown:', 'thumbsDown', {
              value: channel.name,
            }),
          ]),
        ],
        channels,
      ),
    ]).toMatchSnapshot();
  });
});
