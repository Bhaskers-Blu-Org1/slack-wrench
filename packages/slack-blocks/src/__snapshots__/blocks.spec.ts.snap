// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slack Block widgets can compose blocks into bigger things 1`] = `
Array [
  Object {
    "accessory": Object {
      "text": Object {
        "emoji": true,
        "text": "Search",
        "type": "plain_text",
      },
      "type": "button",
      "value": "changeSearch",
    },
    "text": Object {
      "text": "Let me help you find some channels.",
      "type": "mrkdwn",
    },
    "type": "section",
  },
  Object {
    "type": "divider",
  },
  Object {
    "text": Object {
      "text": "*Channels*",
      "type": "mrkdwn",
    },
    "type": "section",
  },
  Object {
    "text": Object {
      "text": "*house-ravenclaw*
Discuss Ravenclaw business",
      "type": "mrkdwn",
    },
    "type": "section",
  },
  Object {
    "elements": Array [
      Object {
        "text": "120 members
Last post: <!date^1575643433^{date_pretty}|1575643433>",
        "type": "mrkdwn",
      },
    ],
    "type": "context",
  },
  Object {
    "elements": Array [
      Object {
        "text": Object {
          "emoji": true,
          "text": ":thumbsup:",
          "type": "plain_text",
        },
        "type": "button",
        "value": "house-ravenclaw",
      },
      Object {
        "text": Object {
          "emoji": true,
          "text": ":thumbsdown:",
          "type": "plain_text",
        },
        "type": "button",
        "value": "house-ravenclaw",
      },
    ],
    "type": "actions",
  },
  Object {
    "text": Object {
      "text": "*house-hufflepuff*
Discuss Hufflepuff puffery",
      "type": "mrkdwn",
    },
    "type": "section",
  },
  Object {
    "elements": Array [
      Object {
        "text": "107 members
Last post: <!date^1575643433^{date_pretty}|1575643433>",
        "type": "mrkdwn",
      },
    ],
    "type": "context",
  },
  Object {
    "elements": Array [
      Object {
        "text": Object {
          "emoji": true,
          "text": ":thumbsup:",
          "type": "plain_text",
        },
        "type": "button",
        "value": "house-hufflepuff",
      },
      Object {
        "text": Object {
          "emoji": true,
          "text": ":thumbsdown:",
          "type": "plain_text",
        },
        "type": "button",
        "value": "house-hufflepuff",
      },
    ],
    "type": "actions",
  },
]
`;

exports[`Slack Block widgets creates a formatted date 1`] = `"<!date^1567191506^Posted {date_num}|Posted 2019-09-30>"`;

exports[`Slack Block widgets renders a button 1`] = `
Object {
  "text": Object {
    "emoji": true,
    "text": "Ravenclaw is obviously the best house.",
    "type": "plain_text",
  },
  "type": "button",
  "value": "ravenAgree",
}
`;

exports[`Slack Block widgets renders a divider 1`] = `
Object {
  "type": "divider",
}
`;

exports[`Slack Block widgets renders a markdown section 1`] = `
Object {
  "text": Object {
    "text": "Ravenclaw is obviously the best house.",
    "type": "mrkdwn",
  },
  "type": "section",
}
`;

exports[`Slack Block widgets renders a plaintext block 1`] = `
Object {
  "emoji": true,
  "text": "Ravenclaw is obviously the best house.",
  "type": "plain_text",
}
`;

exports[`Slack Block widgets renders a plaintext block without emoji 1`] = `
Object {
  "emoji": false,
  "text": "Ravenclaw is obviously the best house.",
  "type": "plain_text",
}
`;

exports[`Slack Block widgets renders a section 1`] = `
Object {
  "accessory": Object {
    "text": Object {
      "emoji": true,
      "text": "Ravenclaw is obviously the best house.",
      "type": "plain_text",
    },
    "type": "button",
    "value": "ravenAgree",
  },
  "fields": Array [
    Object {
      "emoji": true,
      "text": "Ravenclaw is obviously the best house.",
      "type": "plain_text",
    },
    Object {
      "text": "Ravenclaw is obviously the best house.",
      "type": "mrkdwn",
    },
  ],
  "text": Object {
    "emoji": true,
    "text": "Ravenclaw is obviously the best house.",
    "type": "plain_text",
  },
  "type": "section",
}
`;

exports[`Slack Block widgets renders an actions 1`] = `
Object {
  "elements": Array [
    Object {
      "text": Object {
        "emoji": true,
        "text": "Ravenclaw is obviously the best house.",
        "type": "plain_text",
      },
      "type": "button",
      "value": "ravenAgree",
    },
    Object {
      "text": Object {
        "emoji": true,
        "text": "Ravenclaw is obviously the best house.",
        "type": "plain_text",
      },
      "type": "button",
      "value": "ravenAgree-2",
    },
  ],
  "type": "actions",
}
`;

exports[`Slack Block widgets renders context 1`] = `
Object {
  "elements": Array [
    Object {
      "emoji": true,
      "text": "Ravenclaw is obviously the best house.",
      "type": "plain_text",
    },
  ],
  "type": "context",
}
`;

exports[`Slack Block widgets renders markdown text 1`] = `
Object {
  "text": "Ravenclaw is obviously the best house.",
  "type": "mrkdwn",
}
`;
